### PR TITLE
Upgrade to react-native 0.66

### PR DESCRIPTION
**Pull request by Kudo**
# Why

follow up https://github.com/expo/expo/pull/15817#discussion_r781366360
after we use the react-native from npm, the react-native version we used in expo-go is decoupled with other stuffs. that makes us upgrade react-native easier.

# How

1. update versions in package.json
2. sync changes from [react-native-upgrade-helper](https://react-native-community.github.io/upgrade-helper/?from=0.64.2&to=0.66.4)
3. ~`okio` and `okhttp` were be upgraded in react-native 0.66. this is the same to~ https://github.com/expo/expo/pull/15915 landed.
4. `RCTComponentData` changed its designated initializer in react-native 0.65. to make it compatible with older version, i introduced the `EXComponentDataCompatibleWrapper` class to support different versions.
5. [react-native-lab] to fix expo-go build error, we should patch react-native to use the codegen library from its repo but not from node_modules:  https://github.com/expo/react-native/pull/29.

# Test Plan

- CI passed
- bare-expo build & launch test
- unversioned expo-go build & launch test

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-modules-core/ios/NativeModulesProxy/EXComponentDataCompatibleWrapper.m [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                              bridge:(RCTBridge *)bridge
+                     eventDispatcher:(id<RCTEventDispatcherProtocol>)eventDispatcher
+{
+  if ([self respondsToSelector:@selector(initWithManagerClass:bridge:)]) {
+    return [super initWithManagerClass:managerClass bridge:bridge];
+  }
+  return [super initWithManagerClass:managerClass bridge:bridge eventDispatcher:eventDispatcher];
+}
+
+@end
```
Comment (bbarthec) [inline packages/expo-modules-core/ios/NativeModulesProxy/EXComponentDataCompatibleWrapper.m:28]: Very clean solution 👏

**Review by bbarthec — APPROVED** [review]
Superb work! 
Nice job taking core of each edge case detected 🚀 
I've briefly looked at the [upgrade from `rn@0.66.4` to `rn@0.67.0-rc.6`](https://react-native-community.github.io/upgrade-helper/?from=0.66.4&to=0.67.0-rc.6) and observed that there's a change in `android/build.gradle` regarding fetching `react-native` from Maven Central. Should we care about it? Maybe that might cause some additional issues? 🤔

**Review by brentvatne — APPROVED** [review]

**Review by lukmccall — APPROVED** [review]

**Comment by Kudo** [comment]
> Superb work! Nice job taking core of each edge case detected 🚀 I've briefly looked at the [upgrade from `rn@0.66.4` to `rn@0.67.0-rc.6`](https://react-native-community.github.io/upgrade-helper/?from=0.66.4&to=0.67.0-rc.6) and observed that there's a change in `android/build.gradle` regarding fetching `react-native` from Maven Central. Should we care about it? Maybe that might cause some additional issues? 🤔

that is a nice to have notation that prevent users accidentally to fetch [the old react-native from maven central](https://search.maven.org/artifact/com.facebook.react/react-native) when they do not fetch from node_modules. since we have ` implementation "com.facebook.react:react-native:+"`, in this case, the old version on maven central is also a match.
anyway, i'll try further to upgrade 0.67 and include the gradle change.

**Comment by mangkoran** [comment]
Are there any release with this PR included (beta maybe?)?

I want to try out in my project.

**Comment by Kudo** [comment]
@mangkoran you can try to have a react-native 0.66 project and install expo modules
```sh
$ npx react-native init RN066 --version 0.66
$ cd RN066
$ npx install-expo-modules
```

**Comment by mangkoran** [comment]
@Kudo I encountered some issue when trying to run `npx install-expo-modules`. I filed new issue https://github.com/expo/expo-cli/issues/4163.

**Pull request merged**